source/docs/software/wpilib-tools/path-planning/wpilib-trajectories/ramsete.rst

```diff
@@ -0,0 +1,55 @@
+
+.. note:: The "goal pose" represents the position that the robot should be at at a particular timestep when tracking the trajectory. It does NOT represent the final endpoint of the trajectory.
+
+The controller can be updated using the ``Calculate`` (C++) / ``calculate`` (Java) method. There are two overloads for this method. Both of these overloads accept the current robot position as the first parameter. For the other parameters, one of these overloads takes in the goal as three separate parameters (pose, linear velocity, and angular velocity) whereas the other overload accepts a ``Trajectory.State`` object, which contains information about the goal pose. For its ease, users should use the latter method when tracking trajectories.
+
+.. tabs::
+
+   .. code-tab:: java
+
+      var goal = trajectory.sample(3.4); // sample the trajectory at 3.4 seconds from the beginning
+      ChassisSpeeds adjustedSpeeds = controller.calculate(currentRobotPose, goal);
+
+
+   .. code-tab:: c++
+
+      const auto goal = trajectory.Sample(3.4_s); // sample the trajectory at 3.4 seconds from the beginning
+      ChassisSpeeds adjustedSpeeds = controller.Calculate(currentRobotPose, goal);
+
+Using the adjusted velocities
+-----------------------------
+The adjusted velocities are of type ``ChassisSpeeds``, which contains a ``vx`` (linear velocity in the forward direction), a ``vy`` (linear velocity in the sideways direction), and an ``omega`` (angular velocity around the center of the robot frame). Because the Ramsete controller is a controller for non-holonomic robots (robots which cannot move sideways), the adjusted speeds object has a ``vy`` of zero.
+
+The returned adjusted speeds can be converted to usable speeds using the kinematics classes for your drivetrain type. For example, the adjusted velocities can be converted to left and right velocities for a differential drive using a ``DifferentialDriveKinematics`` object.
+
+.. tabs::
+
+   .. code-tab:: java
+
+      ChassisSpeeds adjustedSpeeds = controller.calculate(currentRobotPose, goal);
+      var wheelSpeeds = kinematics.toWheelSpeeds(adjustedSpeeds);
+      double left = wheelSpeeds.leftMetersPerSecond;
+      double right = wheelSpeeds.rightMetersPerSecond;
+
+
+   .. code-tab:: cpp
+
+      ChassisSpeeds adjustedSpeeds = controller.Calculate(currentRobotPose, goal);
+      auto wheelSpeeds = kinematics.ToWheelSpeeds(adjustedSpeeds);
+      units::meter_per_second_t left = wheelSpeeds.left;
+      units::meter_per_second_t right = wheelSpeeds.right;
+
+Because these new left and right velocities are still speeds and not voltages, two PID Controllers, one for each side may be used to track these velocities. Either the WPILib PID Controller can be used, or the Velocity PID feature on smart motor controllers such as the TalonSRX and the SparkMAX can be used.
```

Comment:
So to make sure I'm understanding this correctly, you'd do PID on the velocity, with the left and right wheel outputs?

Reply:
Yes, you would run one PID controller per side

Reply:
That seems like there could be some inconsistency, especially in tuning your PID controller to follow this.

Reply:
That's simply the nature of the controller. Teams can use whatever reference controller they want to "follow" the left and right velocities -- PID (with a feedforward) is the simplest and most effective way to do this -- and it works well in practice.